tools/busychat/index.js

```diff
@@ -0,0 +1,419 @@
+      this._joinChannel(channel)
+    } else if (trimmed.startsWith('/part')) {
+      const channel = trimmed.substr(5).trim()
+      this._partChannel(channel)
+    } else if (trimmed.startsWith('/status')) {
+      let out = ['channels:']
+      for (let c of this._channels) {
+        out.push(' - [' + c + ']')
+      }
+      this._display(out.join('\n'))
+    } else if (trimmed.startsWith('/list')) {
+      this._display('sorry... I don\'t know how to list channels yet')
+    } else {
+      // publish a message to the active channel
+      this._publishMessage(this._activeChannel, value)
+    }
+  }
+
+  _displayHelp () {
+    this._display(`
+busychat n3h test app
+/help                - display this help
+/name [new_name]     - give yourself a new name
+/join [channel_name] - listen to [channel_name] - create if it doesn't exist
+                       messages will send to the most recently joined channel
+                       joining an already joined channel will mark it active
+/part [channel_name] - leave, stop listening to [channel_name]
+/list                - list all known channels
+/status              - show info such as which channels you have joined
+/quit                - exit busychat
+`)
+  }
+
+  _ipcSend (data) {
+    let msg = Buffer.from(JSON.stringify(data), 'utf8')
+    msg = msgpack.encode({ name: 'json', data: msg }).toString('base64')
+    this._ipc.send(Array.from(this._ipc.keys()), msg)
+  }
+
+  _handleIpcEvent (e) {
+    if (e.type === 'message') {
+      const outer = msgpack.decode(Buffer.from(e.buffer, 'base64'))
+      const name = outer.name.toString()
+      if (name === 'json') {
+        const data = JSON.parse(outer.data.toString('utf8'))
+        if (typeof data.method === 'string') {
+          this._handleParsedIpcEvent(data)
+        }
+      }
+    }
+  }
+
+  _handleParsedIpcEvent (e) {
+    switch (e.method) {
+      case 'peerConnected':
+        this._display(JSON.stringify(e))
+        break
+      case 'handleGetPublishingEntryList':
+      case 'handleGetHoldingEntryList':
+      case 'handleGetPublishingMetaList':
+      case 'handleGetHoldingMetaList':
+        // ignore some explicitly un-handled messages
+        break
+      case 'handleStoreEntry':
+        this._stats.entryCount += 1
+        this._updatePrompt()
+
+        this._fetchMessage(e.dnaAddress, e.address)
+
+        // if (!this._channels.has(e.dnaAddress)) {
+        //   return
+        // }
+        // this._display(
+        //   '[' + e.dnaAddress + ']', e.content.from + ':', e.content.data)
+        break
+      case 'handleStoreMeta':
+        this._fetchMeta(e.dnaAddress, e.entryAddress, e.attribute)
+        break
+      case 'fetchEntryResult':
+        if (!this._channels.has(e.dnaAddress)) {
+          return
+        }
+
+        this._display(
+          '[' + e.dnaAddress + ']', e.content.from + ':', e.content.data)
+        break
+      case 'fetchMetaResult':
+        // console.log(e)
```

Comment:
Maybe use meta as just a 'reply to' link to another entry.

Reply:
Oo, I like that idea. Yeah, it's kind of a fun challenge to implement something that is functional while also mimicing the patterns of holochain